beacon-chain/operations/attestation.go

```diff
@@ -0,0 +1,166 @@
+// Handler defines an interface for a struct equipped for receiving block operations.
+type Handler interface {
+	HandleAttestation(context.Context, proto.Message) error
+}
+
+// retrieves a lock for the specific data root.
+func (s *Service) retrieveLock(key [32]byte) *sync.Mutex {
+	keyString := string(key[:])
+	mutex := &sync.Mutex{}
+	item := s.attestationLockCache.Get(keyString)
+	if item == nil {
+		s.attestationLockCache.Set(keyString, mutex, 5*time.Minute)
+		return mutex
+	}
+	if item.Expired() {
+		s.attestationLockCache.Set(keyString, mutex, 5*time.Minute)
+		item.Release()
+		return mutex
+	}
+	return item.Value().(*sync.Mutex)
+}
+
+// AttestationPool returns the attestations that have not seen on the beacon chain,
+// the attestations are returned in target epoch ascending order and up to MaxAttestations
+// capacity. The attestations returned will be verified against the head state up to requested slot.
+// When fails attestation, the attestation will be removed from the pool.
+func (s *Service) AttestationPool(ctx context.Context, requestedSlot uint64) ([]*ethpb.Attestation, error) {
+	s.attestationPoolLock.Lock()
+	defer s.attestationPoolLock.Unlock()
+
+	ctx, span := trace.StartSpan(ctx, "operations.AttestationPool")
+	defer span.End()
+
+	atts := make([]*ethpb.Attestation, 0, len(s.attestationPool))
+
+	bState, err := s.beaconDB.HeadState(ctx)
+	if err != nil {
+		return nil, errors.New("could not retrieve attestations from DB")
+	}
+
+	if bState.Slot < requestedSlot {
+		bState, err = state.ProcessSlots(ctx, bState, requestedSlot)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not process slots up to %d", requestedSlot)
+		}
+	}
+
+	var validAttsCount uint64
+	for root, ac := range s.attestationPool {
+		for _, att := range ac.ToAttestations() {
+			if s.recentAttestationBitlist.Contains(root, att.AggregationBits) {
+				continue
+			}
+
+			if _, err = blocks.ProcessAttestation(ctx, bState, att); err != nil {
+				delete(s.attestationPool, root)
+				continue
+			}
+
+			validAttsCount++
+			// Stop the max attestation number per beacon block is reached.
+			if validAttsCount == params.BeaconConfig().MaxAttestations {
+				break
+			}
+
+			atts = append(atts, att)
+		}
+	}
+
+	return atts, nil
+}
+
+// AttestationPoolNoVerify returns every attestation from the attestation pool.
+func (s *Service) AttestationPoolNoVerify(ctx context.Context) ([]*ethpb.Attestation, error) {
+	s.attestationPoolLock.Lock()
+	defer s.attestationPoolLock.Unlock()
+
+	atts := make([]*ethpb.Attestation, 0, len(s.attestationPool))
+
+	for _, ac := range s.attestationPool {
+		atts = append(atts, ac.ToAttestations()...)
+	}
+
+	return atts, nil
+}
+
+// HandleAttestation processes a received attestation message.
+func (s *Service) HandleAttestation(ctx context.Context, message proto.Message) error {
+	s.attestationPoolLock.Lock()
+	defer s.attestationPoolLock.Unlock()
+
+	ctx, span := trace.StartSpan(ctx, "operations.HandleAttestation")
```

Comment:
Please put this at the start of the method. It would be nice to see an locks blocking as part of this span.